### ISSUE TITLE
Common Game functions demanding the speed of C

### ISSUE THREAD

**Issue by mcpalmer1980** [pygame/pygame#1152]
Following a somewhat length introduction I will layout my ideas for a large pygame2 add-on that will perform common game operations that are too slow for python to run in compiled C code instead. This will move Pygame from hobbyist framework to usable game engine territory.

I am nearly finished with my third solo pygame project since I began experimenting almost a year ago. Python and pygame present a work-flow I prefer over the Godot engine that I originally planned to use for Flyboy, not to mention lower CPU and memory usage despite the blitting bottlenecks of sdl1.2 surfaces. My games all run at 8% CPU with a 30fps lock and a 1600x900 resolution with pixel alpha on all sprites above the tile layer. A visit to mcpalmer1980.itch.io should exhibit my proficiency in game design and the pygame framework. Just don't look at my menu code because I wrote it before gaining a firm grasp on Python.

The upcoming release of Pygame2 truly excites me, even if Android support is still a far way off. I have read several threads regarding how to present the SDL2 rendering interface, and despite many common opinions, I still think a transparent migration could have worked something like this: Blitting surfaces batches textures instead, display.flip() renders the batch, and pixel operations on GPU textures are silently ignored. This would allow existing projects to run unaltered until their developers replace pixel effects with shader effects.

I have looked at PyTMX and PyScroll and like what I see. PyScroll performs quite well in pure python, although not as fast as my own much simpler tilemap class(uses one large buffer and blits the entire window area in one call). I imagine it will perform extremely well using pygame2's hardware texture rendering and it probably doesn't need to be implemented in C. I only mention it because there are grid operations that MUST BE implemented in C, so even if no part of the tilemap rendering is in C, some sort of common grid format must be decided on.

Common operations we need at C speed:

RayCasting: the raycasting code in my tilemap class is 47 lines long. It does the job well enough because my AI only checks once or twice per second.

	def CheckBlocked(self, sx, sy, ex,ey, blocks, sight = 0):
		sx = int(sx / self.cell_width)
		sy = int(sy / self.cell_height)
		ex = int(ex / self.cell_width)
		ey = int(ey / self.cell_height)
		dx = float(ex - sx)
		dy = float(ey - sy)
		if sight and Vector2(dx,dy).GetMagnitude() > sight:
			return True
		if type(blocks)==str:
			blocks = self.filters.get(blocks, blocks)

		b=0
		if dx:
			if sx < ex:
				st = 1
			elif sx > ex:
				st = -1

			dy = dy / abs(dx)
			y = sy-dy

			for x in range(sx, int(ex), st):
				y += dy
				g = self.grid[int(x)][int(y)]
				if not blocks[g]:
					b = 1
					break
			if b:
				return b

		dy = ey - sy
		if dy:
			if sy < ey:
				st = 1
			if sy > ey:
				st = -1

			dx = dx / abs(dy)
			x = sx - dx

			for y in range(sy, int(ey), st):
				x += dx
				g = self.grid[int(x)][int(y)]
				if not blocks[g]:
					b = 1
					break
		return b

Pathfinding: I use A* path-finding in Alien Park, my latest project soon to appear on itch.io. It takes 3-10ms to run, so I must use it VERY SPARINGLY. Chasers check a cache of the last 30 tiles the player moved across and try to follow that if possible, reverting to my simplified A* implementation when it cannot. My implementation doesn't find the best path because it stops the moment any path is found instead.

Grid Collisions: thusfar I've used sprites equal to the cell size in a platformer, and point collision based on a character's feet in top-down maps. In my platformer I looked up the cell at the grid coordinates of his head and feet at his right or left size depending on his direction of travel. With larger characters and/or smaller cells it becomes necessary to determine every cell a rect overlaps, and then determine if any are solid or triggers. Some part of this could be accelerated if other grid operations are already implimented in C.

Detailed Collision Detection: I frequently use rect.collidelistall in my games. I have not played with the pygame.Mask yet, feeling no need for pixel perfection. But I've always had trouble determining which edge of a rect has collided with the other. I fudged it in Flyboy by using the degree of overlap, but with moving platforms that players can ride, bump, or be squashed by, my method is just barely adequate. Something like this would be better.

	rect.collide_vector(other_rect, my_velocity): return (TOP, BOTTOM, LEFT, or RIGHT), ?force / remainder?

	>here is the code I settled for<

	def Collisions(self):
		if self.was_on and not self.on_floor and self.Act in (self.Standing, self.WalkLeft, self.WalkRight, self.Looking):
			self.SetPos(self.x, self.was_on.rect.top - self.from_floor)

		self.was_on = self.on_robo
		self.contact = None
		self.on_robo = None
		self.contacts = 0
		foot = (self.x + self.oCenter, self.y + self.oFoot)
		robots = self.game.robots()
		self.tHead = None
		self.tLeft = None
		self.tRight = None

		collisions = self.rect.collidelistall(robots)
		if collisions == []:
			return

		self.contact_list = []
		for i in collisions:
			self.contacts += 1
			r = robots[i]
			self.contact_list.append(r)
			self.contact = r
			ol = self.rect.clip(r.rect)
			hit = False
			rect = self.rect
			if not r.solid:
				continue

			if ol.top == r.rect.top and ol.height < rect.height / 3:
				self.on_floor = ROBO
				self.on_robo = r
				self.was_on = r

			elif ol.left == r.rect.left:
				self.tRight = r
			elif ol.right == r.rect.right:
				self.tLeft = r
			elif ol.bottom == r.rect.bottom and ol.height < rect.height / 3:
				self.tHead = r
		[ self.Hit(r) for r in self.contact_list if r.freeze < 1 ]

**Comment by illume** [comment]
Hello!

This sounds like a good list. Shall we discuss one at a time?

Can you start a new issue for pathfinding (or another one if you prefer), and we can discuss it there? 

ps. Have you seen this post? http://renesd.blogspot.com/2017/03/four-new-pygame-things-for-slow.html  Two of these things have been done, and the sprite-in-cython one was started by @dlon already.

**Comment by mcpalmer1980** [comment]
My internet access is extremely limited to 30 minute sessions a couple times a week. I apologize for not following common convections and hope you might forgive my failures and try to work within my limitations.

**Comment by illume** [comment]
Ok. Well, then it seems we need a good understanding of common structures for tilemap data.

Is one structure able to work well? Or are there a couple of them which maybe we need to work with? Are there reasons for one structure or the other?

**Comment by mcpalmer1980** [comment]
There are several ways to represent a tilemap in memory. Lets assume that all tilemaps will be a grid of squares, whether drawn flat or isometrically. Three dimensional isometric maps are a difficult subject, as are hexagonal tilemaps. I doubt many pygame users would tread those waters, and if they do, they'll need to implement their own collision, ray-casting, and path-finding algorithms.

The designer of PyScroll prides himself with the handling of tiles as integers instead of python objects with their 25k of overhead(his words, I'm not sure about accuracy). I also use integers for my tilemap class, which doesn't support layered drawing and limitless map size the way pyScroll does. I use a single buffer surface that I decided to limit around 4032x2688 @ 4bytes/pixel(43.352mb). PyScroll, by blitting cells and layers individually while looping through the grid, uses 3-4 times the CPU time as mine, but probably 10% the memory. But I digress, the topic is data, not rendering.

**Things needing consideration**
- some cells will be solid, blocking movement onto them and sight through them (walls, doors, etc)
- some cells will be impassible, blocking movement onto them but not sight through them(water, holes, etc)
- some cells will be hazards, blocking neither sight nor movement, but harming or killing the player (holes, spikes)
- most hazards will but considered solid by enemies so they don't move onto them(water, holes, spikes)
- some cells will be partially blocked, considered blocked by path-finding but not by collisions detection (slanted floors, edges and corners of terrain cells)
- partially blocked cells might include a polygon or other shape attribute for collision detection, but ignored by path-finding and ray-casting
- the grid must be stored as some sort of 2-dimensional array, either a list of lists, a numpy array, or perhaps a new python.math.grid
- how common is numpy among pygame users?
- how much speed benefit would numpy offer versus its size as a dependency?
- how much would numpy inflate the size of frozen binaries?
- physics engines and game engines such as Godot often create one large collision shape for the tilemap
- looking up the grid location like in the old days is simpler and less intensive
- physics engines already exist for python and are neither newbie nor resource friendly
- adding ray-casting and path-finding to pygame may require an entire tilemap class
- such a class may need to be an add-on
- pyTMX already loads Tiled tilemaps for use with the PyScroll library
- perhaps the PyScroll project can be extended with some low-level object code for pathfinding/raycasting/etc

**Data structures to store grids**

**method 1: use an existing integer array**
My tilemap class uses this method. The tilemap is a list of lists. I add a second layer for additional information but the primary layer determines if the layer is solid or not. I use lists such as isSOLID, playerCANSEE, enemyCANSTEP, each with the cell numbers that apply. Usage looks something like this.

```
GRASS=1
WALL=2
HOLE=3
FENCES=[4,5,6,7]
CELL_TYPE_COUNT = 8

isSOLID = [WALL]+FENCES
enemySEE = [GRASS, HOLE]

# MakeFilter will make an array of boolean values for all cell types because I think that
# 'if canSTEP[cell_value]' runs quicker than 'if cell_value in canSTEP'
enemy.canSEE = tilemap.MakeFilter(enemySEE, CELL_TYPE_COUNT)

if not tilemap.CheckBlocked(enemy.gridx, enemy.gridy, player.gridx, player.gridy, enemy.canSEE):
	enemy.ShootAt(player)
elif enemy.IsItTimeToSearch():
	path = tilemap.FindPath(enemy.gridx, enemy.gridy, player.gridx, player.gridy, enemy.canSTEP)
	if path:
		enemy.FollowPath(path)

```
PROS:
uses no additional memory
quick

CONS:
cannot store additional information about cell status
requires many cell types
additional layers cannot effect whether the cell is solid or not


**method 2: use multiple exiting integer arrays**
basically the same as method 1, except that algorithms need to check each grid layer to see if the object there is solid or not. Anyone who's unsure why there needs to be multiple tilemap layers can look at the pyscroll demo. That demo uses several layers to draw a cool map.

PROS:
uses little memory
additional layers can effect whether the cell is solid or not

CONS:
cannot store much additional information about cell status
slower than method 1

**Method 3: Create a new boolean collision array**
While loading the tilemap data, create a boolean array set to 1 if any objects at the grid location make it solid, or 0 if not.

PROS:
probably fastest method at runtime
additional layers can effect whether the cell is solid or not
Uses only slightly more memory than method 1(one byte per 8 grid squares)

CONS:
the collision array must be updated each time the contents of a grid square changes
cannot have different sight or traversal results among players and various enemy types...
it would take extra code to prevent enemies from falling in water or allowing them to fly above bushes
cannot store much additional information about cell status


**Method 4: create a new byte collision array**
While loading the tilemap data, create a byte array using up to 8 bit masks. Pass the bit mask to collision routines to determine which to look up.
```
MASK1 = (WALL, WATER, FENCE)
MASK2 = (WALL, FENCE)
MASKS = (MASK1, MASK2)
tilemap.SetMasks(MASKS)

if not CheckBlocked(enemy.gridx, enemy.gridy, target.gridx, target.gridy, MASK1):
	go there
```
PROS:
nearly as fast as method 3
additional layers can effect whether the cell is solid or not
can have different sight or traversal results among players and various enemy types

CONS:
the collision array must be updated each time the contents of a grid square changes
cannot store much additional information about cell status


**Method 5: use the integer array as the key to an object lookup table**
The integer values in the array are used to look up the object or dictionary for each cell.
```
GRASS = Cell(solid=False, canstep = True, hazard=False) or GRASS={canstep: True}
WATER = Cell(solid=False, canstep = False, hazard=True) or WATER={hazard: True}
WALL = Cell(solid=True, canstep=False, hazard=False) or WALL={solid: True}
tilemap.SetCells((GRASS, WATER, HOLE, ETC))
```
PROS:
lots of options for attribute types
can avoid multiple layers
grid uses no extra memory

CONS:
still cannot set properties for each individual cell


**Method 6: use one array of objects for everything**

This is probably the easiest to use and most powerful method, but with memory drawback on large maps. The grid is made up of object instances or dictionaries that include mandatory collision data and any other attributes desired. For instance, a door can have a key_number value, a trigger plate could have an event_id value, a torch could have a time_left_burning value, etc. For Flyboy(mcpalmer1980.itch.io/Flyboy) I stored such information in event and attribute lists that could be looked up in dictionaries like: event = mapdb.events.get( (gridx, gridy), None )

I'm not sure if there's any need for individual cells to have methods, or what performance differences exist between objects and dictionaries. I will used dictionaries in the following example.
```
#place a new invisible key on the grid, ignoring its current contents
new_key = objects['key'].copy()
new_key['key_id'] = 2
new_key['visible'] = False
tilemap.SetCell(2, 4, new_key)
```
PROS:
unlimited options for object attributes
individual cells can be made solid, invisible, hazardous, etc
can avoid multiple layers

CONS:
grid can eat up lots of memory

**Comment by robertpfeiffer** [comment]
A lot of these may be better solved by implementing better data structures (like quad trees) or better algorithms (like jump point search) in pure python, instead of implementing slow ones in C.

You can use numpy right now together with pygame, and I see no reason to always add a dependency on numpy to pygame, instead of letting game devs decide on a case by case basis.
As for grid collisions: Does it matter how many tiles it overlaps during movement? I would assume on right movement (without loss of generality) you could stop on collision with the leftmost colliding tile.

I don't think anything in here demands "the speed of C", in the sense that it needs to be written in C and added to pygame itself.

I am (not publically) working on a modernGL-based drop-in replacement for psycroll in which tile layers are rendered as single quads, with tile lookup done in the pixel shader. I have also written a simple particle and sprite atlas/sprite batching system. It's only a couple hundred lines. If you are really that strapped for performance and don't need pixel drawing primitives, you can use openGL right now, or go for a hybrid approach where occasionally you use pygame to draw onto some surfaces and copy them to GPU textures, while using GL for most things. It is not an either-or thing.

Adding a portable, SDL-render based tilemap renderer to pygame might be a good idea, but it might just be just a renderer for a single tilemap layer, and we could build a SpriteGroup interface on top of it in Python.

**Comment by mcpalmer1980** [comment]
As I said before, A* takes 3-10ms for pathfinding. Also, lighting and AI behavior demands lots of raycasting. Particle effects that interact with the grid would be to slow with lots of particles. A quad tree algo will not solve these problems written in pure python.

I am thinking of using pytmx and a renderer-based pyscroll as a basis for the above tilemap features once the texture rendering API is finalized.

**Comment by robertpfeiffer** [comment]
We can write a simple tilemap blitter in C that stores tilemaps in an (array)[https://docs.python.org/3/library/array.html] and takes a tilesheet surface. This could be a method for pygame.Surface that gets wrapped in a nice pythonic API so you don't have to supply the shape of the tilemap and tile size every time.

Something like `surface.tile_render(tilemap, tilesheet_surface, tile_size, source_rect, dest_rect)`. But we'd have to benchmark whether that would be worth it.

**Comment by mcpalmer1980** [comment]
I wouldn't worry about the tilemap blitter. The more I think about it the less sense it makes to provide a generic blitter that won't fill many users' purposes. I thought  of instead how I would go about drawing my own new tilemap class and what drawing features I would need to speed it up.

My answer can be found in a new topic about adding transform_and_draw to the image or sprite class, which would apply transforms for the specific image as well as a global transform vector that would apply to the x/y coordinates of all images that reference it.

I've already added pygame._sdl2.video style image to the pytmx loader and my long term goal is to port pyscroll to use pygame2 textures once the api is finalized and documented. Then I will add raycasting, pathfinding, lighting, etc to that(probably in C)

**Comment by robertpfeiffer** [comment]
I wrote a tilemap blitter in C just to see how fast it could get. I got a 2x increase in FPS on a simple game with keyboard input. This makes the C impl based on an array tilemap and a tilesheet surface take 20%-40% as much time as a naive Python version.

With SDL2 SDL_renderer, I wouldn't expect such speed-ups unless we are willing to re-write the tilemap renderer for every backend. Fortunately, rendering subtextures of the same parent texture is already optimised with the SDL2 GL backend, but not as much as sprite batching in XNA or libGDX are. The fastest way would still be to write a shader-based tilemap renderer, but SDL2 should get you far enough.

As for terrain collision, path finding, and visibility, do you agree that these things should not necessarily be handled by the same data structures as tilemap rendering?

**Comment by mcpalmer1980** [comment]
Yes, I agree. Convenience might prefer them together, but in practice, I have used a separate grid for pathfinding and such(not so much for collision detections).

**Comment by mcpalmer1980** [comment]
I've been giving more thought to your idea of adding a simple tilemap render to pygame. Initially I reacted negativiely to this idea because I believed my transform_and_draw suggestion would allow tilemaps and particle systems to be built with ease. The more I think of it the more sense your suggestion makes.

Adding a generic transform_and_draw method would take far more consideration and coding than a simple tilemap renderer. The two main uses for my transform_and_draw method would be for tilemaps and for particles. Of the two, a simple particle system would be easy to impliment in python but a tilemap renderer would not. All pygame users would benefit from tilemaps, but many would not be able to impliment them properly. Looping through onscreen cells is too slow, whereas rendering tiles into more managable quadrants is too complicated.

So a method that takes a 2d list/array of subsurface or texture references and renders it would be great. Users add add more layers by calling it again. Such a feature isn't a set milestone for 2.0, but maybe it should be. Adding backwards compatable SDL2 support to pygame is a worthy goal and will be a great achivement, but perhaps built in tilemap rendering could be that one feature to bring pygame back to its former glory.

There are some considerations to make.

- Will it support surfaces or textures only
- Will it take a 2d array of texture references or indexes into a texture list
- How should the grid be culled:
-pass a upper-left coordinate
-pass a center coordinate
-pass a rectangular region


```
grid = pickle.load('my_map.dat')
tset = pygame.tiles.Tileset('set.png', columns, rows)
tmap = pygame.tiles.Tilemap(tset, grid)
tmap.set_scale(1.5)
tmap.centerx = 500
tmap.render(renderer)
tmap.blit(surface, (0,0))
```

or as you said earlier

`surface.tile_render(tilemap, tilesheet_surface, tile_size, source_rect, dest_rect)`

On second thought, even the simplest tilemap renderer demands quite a lot of consideration(probably more time than actually implimenting the thing). Oh well, may as well post this anyway.

**Comment by illume** [comment]
TLDR; tilemap optimizations are a good idea. We need games to benchmark.

- A tile-map renderer in C can be 2x faster is confirmed.
- the _sprite cython implementation by dlon also showed a 2x speedup.
- tile-map renderers in python exist that are also faster (than pyscroll) by N(unclear) as well.



What we have to benchmark:
- do we have a list of games that use tile renderers?
- pyscroll (what games use this we can use to test?)
- The generic "groups of images" blitter we have I guess is the sprite groups. Note, they also handle layers, and dirty rects. 
- We also have Surface.blits(). Has this been used?
- consideration should be given to the slow configuration (no hardware acceleration).

Some other things to explore:
- No changes preferred: If we can come up with a speedup which improves existing games/libraries APIs with no changes, that is a double plus good result. (But maybe we can't speed things up with existing functions)
- Can we give a function to pyscroll (and other games) they can use instead to speed things up.

**Comment by MyreMylar** [comment]
- [ ] Split this issue into separate feature requests
